### DISCUSSION TITLE
completion

### DISCUSSION THREAD

**Discussion by aloispichler** [hrsh7th/nvim-cmp#1740] [Q&A]
I enjoy the following completion behavior from the Julia-language server.
Assume that `\alpha` is a snippet which inserts `α`:
1. While typing `\al…`, the dropdown is offered;
2. Tabbing to `\alpha` inserts `α` immediately in the text.
3. Just continue typing; it is not necessary to confirm the selection.

How can this behavior be achieved with nvim-cmp? – Many thanks!

**Comment by fitrh** [comment]
You can try bind `<Tab>` to `cmp.select_next_item({ behavior = cmp.SelectBehavior.Insert })`

**Reply by aloispichler** [reply]
Thanks. what I tried is
```
cmp.setup {
  snippet = {
    expand = function(args)
      luasnip.lsp_expand(args.body)
    end,
  },
  mapping = cmp.mapping.preset.insert {
    ['<C-b>'] = cmp.mapping.scroll_docs(-4),
    ['<C-f>'] = cmp.mapping.scroll_docs(4),
    ['<C-Space>'] = cmp.mapping.complete(),
    ['<C-e>'] = cmp.mapping.abort(),
    ['<CR>'] = cmp.mapping.confirm({  -- <TAB>
      -- behavior = cmp.ConfirmBehavior.Replace,  -- if active, replaces succeeding text
      select = true,
    }),
    ['<Tab>'] = cmp.mapping(function(fallback)
      if cmp.visible() then
        cmp.select_next_item({behavior = cmp.SelectBehavior.Insert})
      elseif luasnip.expand_or_jumpable() then
        luasnip.expand_or_jump()
      else
        fallback()
      end
    end, { 'i', 's' }),
    ['<S-Tab>'] = cmp.mapping(function(fallback)
      if cmp.visible() then
        cmp.select_prev_item({behavior = cmp.SelectBehavior.Insert})
      elseif luasnip.jumpable(-1) then
        luasnip.jump(-1)
      else
        fallback()
      end
    end, { 'i', 's' }),
  },
  sources = {
    { name = 'luasnip' },
    { name = 'nvim_lsp' },
  },
}
```
It did not work this way. I must be missing something…

**Reply by fitrh** [reply]
I think if it is a snippet, you need to map it to `cmp.confirm` since `cmp.select_next_item` doesn't expand the snppet

```lua
['<Tab>'] = cmp.mapping(function(fallback)
  if cmp.visible() then
    cmp.confirm({
      select = true,
      behavior = cmp.ConfirmBehavior.Insert,
    })
  elseif luasnip.expand_or_jumpable() then
    luasnip.expand_or_jump()
  else
    fallback()
  end
end, { 'i', 's' }),
```

**Reply by aloispichler** [reply]
Thanks. This code inserts the first snippet in the dropdown, which is actually more than I had :-)

- In the JuliaLSP, I can tab through the dropbown list, each snippet is expanded immediately. The snippet is removed, when tabbing to the next in the dropdown, the next is expanded, etc.
- At this stage, just continue typing anything – no confirmation necessary, the extended, desired snippet is already there.

This is what I’d like to achieve finally with other snippets as well…
